### PR TITLE
Use a stable SARIF identifier

### PR DESCRIPTION
**Pull request by simon-engledew**
:wave: Hello - I am from the GitHub code scanning team! :bow:

We have noticed that your tool is currently generating unstable sarif identifiers. 🕵️ This is against the [SARIF specification](https://docs.oasis-open.org/sarif/sarif/v2.1.0/csprd01/sarif-v2.1.0-csprd01.html#_Ref513193500) 😱.

Unstable identifiers result in suboptimal experience for users of GitHub Code Scanning:

* Users are not able to easily group similar results (for example results for the same CVE in different images)
* Users find results that have been dismissed reappear if the image name or tag changes
* We have a hard limit of 500k identifiers per tool, beyond this point it is not possible to enumerate them all

I've proposed a possible fix that would make your sarif identifier stable. Once you have found something that works for you I can migrate your existing rules. :+1:

### CODE REVIEW

**Comment by afdesk** [comment]
Hi @simon-engledew!
thanks for your feedback! 

I'm studying this issue right now. 
the first moment: this PR changes only the driver's rule ids, and doesn't affect on results's `ruleId`.

**Comment by simon-engledew** [comment]
Hi @afdesk ! 👋 

> this PR changes only the driver's rule ids, and doesn't affect on results's ruleId.

Yes, quite right! Sorry, I meant this as an illustration of how the issue could be resolved - not a complete solution. 🙇

If you are happy with this direction I can spend some more time trying to get the PR into a mergable state?

**Comment by afdesk** [comment]
@simon-engledew thanks a lot for your effort!
it looks good and I don't see problems with these changes.

**Comment by afdesk** [comment]
I'd add more information about scan into `shortDescription`.

**Comment by afdesk** [comment]
@simon-engledew I found a problem. 
simple `VulnerabilityID` may not be unique for scan.

**Comment by afdesk** [comment]
I've created next folder `j`:
```
./j
├── commons-io-1.4.jar
└── t
    └── commons-io-1.4.jar
```
and `trivy` creates the report with not unique ids.

**Comment by simon-engledew** [comment]
@afdesk @krol3

Hi folks :wave: Any progress?

**Comment by simar7** [comment]
Hi @simon-engledew, I was just added to this thread to assist. We'll take a look at it and follow up. 

cc @rahul2393

**Comment by simar7** [comment]
hi again @simon-engledew, this looks good to me. Just one thing, if you could update the tests to reflect this behaviour then we should be able to merge it in. 

I tried to push to you branch but it seems I don't have access. But here's the diff:

```diff
commit 1bb2951145d9630d5069d929a9797e6183b58705
Author: Simon Engledew <simon-engledew@github.com>
Date:   Fri Sep 10 17:16:52 2021 +0100

    Use a stable SARIF identifier
    
    Co-authoered-by: Simar <simar@linux.com>
    Signed-off-by: Simar <simar@linux.com>

diff --git a/contrib/sarif.tpl b/contrib/sarif.tpl
index 1f7bc27..5ecb09e 100644
--- a/contrib/sarif.tpl
+++ b/contrib/sarif.tpl
@@ -20,7 +20,7 @@
                 ,
               {{- end }}
             {
-              "id": {{ printf "%s: %s-%s %s" $result.Target .PkgName .InstalledVersion .VulnerabilityID | toJson }},
+              "id": {{ .VulnerabilityID | toJson }},
               "name": "{{ toSarifRuleName $vulnerabilityType }}",
               "shortDescription": {
                 "text": {{ printf "%v Package: %v" .VulnerabilityID .PkgName | printf "%q" }}
@@ -92,4 +92,4 @@
       }
     }
   ]
-}
\ No newline at end of file
+}
diff --git a/integration/testdata/alpine-310.sarif.golden b/integration/testdata/alpine-310.sarif.golden
index 5698b0f..aa54385 100644
--- a/integration/testdata/alpine-310.sarif.golden
+++ b/integration/testdata/alpine-310.sarif.golden
@@ -11,7 +11,7 @@
           "version": "0.15.0",
           "rules": [
             {
-              "id": "testdata/fixtures/images/alpine-310.tar.gz (alpine 3.10.2): libcrypto1.1-1.1.1c-r0 CVE-2019-1549",
+              "id": "CVE-2019-1549",
               "name": "OS Package Vulnerability (Alpine)",
               "shortDescription": {
                 "text": "CVE-2019-1549 Package: libcrypto1.1"
@@ -37,7 +37,7 @@
               }
             },
             {
-              "id": "testdata/fixtures/images/alpine-310.tar.gz (alpine 3.10.2): libcrypto1.1-1.1.1c-r0 CVE-2019-1551",
+              "id": "CVE-2019-1551",
               "name": "OS Package Vulnerability (Alpine)",
               "shortDescription": {
                 "text": "CVE-2019-1551 Package: libcrypto1.1"
@@ -63,7 +63,7 @@
               }
             },
             {
-              "id": "testdata/fixtures/images/alpine-310.tar.gz (alpine 3.10.2): libcrypto1.1-1.1.1c-r0 CVE-2019-1563",
+              "id": "CVE-2019-1563",
               "name": "OS Package Vulnerability (Alpine)",
               "shortDescription": {
                 "text": "CVE-2019-1563 Package: libcrypto1.1"
@@ -89,7 +89,7 @@
               }
             },
             {
-              "id": "testdata/fixtures/images/alpine-310.tar.gz (alpine 3.10.2): libcrypto1.1-1.1.1c-r0 CVE-2019-1547",
+              "id": "CVE-2019-1547",
               "name": "OS Package Vulnerability (Alpine)",
               "shortDescription": {
                 "text": "CVE-2019-1547 Package: libcrypto1.1"
@@ -115,7 +115,7 @@
               }
             },
             {
-              "id": "testdata/fixtures/images/alpine-310.tar.gz (alpine 3.10.2): libssl1.1-1.1.1c-r0 CVE-2019-1549",
+              "id": "CVE-2019-1549",
               "name": "OS Package Vulnerability (Alpine)",
               "shortDescription": {
                 "text": "CVE-2019-1549 Package: libssl1.1"
@@ -141,7 +141,7 @@
               }
             },
             {
-              "id": "testdata/fixtures/images/alpine-310.tar.gz (alpine 3.10.2): libssl1.1-1.1.1c-r0 CVE-2019-1551",
+              "id": "CVE-2019-1551",
               "name": "OS Package Vulnerability (Alpine)",
               "shortDescription": {
                 "text": "CVE-2019-1551 Package: libssl1.1"
@@ -167,7 +167,7 @@
               }
             },
             {
-              "id": "testdata/fixtures/images/alpine-310.tar.gz (alpine 3.10.2): libssl1.1-1.1.1c-r0 CVE-2019-1563",
+              "id": "CVE-2019-1563",
               "name": "OS Package Vulnerability (Alpine)",
               "shortDescription": {
                 "text": "CVE-2019-1563 Package: libssl1.1"
@@ -193,7 +193,7 @@
               }
             },
             {
-              "id": "testdata/fixtures/images/alpine-310.tar.gz (alpine 3.10.2): libssl1.1-1.1.1c-r0 CVE-2019-1547",
+              "id": "CVE-2019-1547",
               "name": "OS Package Vulnerability (Alpine)",
               "shortDescription": {
                 "text": "CVE-2019-1547 Package: libssl1.1"
```

**Comment by simar7** [comment]
And also if you could sign the CLA that would be very appreciated.

**Comment by simon-engledew** [comment]
@simar7

It's not quite there. AS @afdesk pointed out, the results needed updating.

I've done that, but the next problem is that the rules need to be made unique by `VulnerabilityID` and the `ruleIndex` needs to also reflect that. I'm not sure if that can be achieved using the template language you are using?

**Comment by simon-engledew** [comment]
> And also if you could sign the CLA that would be very appreciated.

I did. 😞 I'll try again.

**Comment by afdesk** [comment]
Hi @simon-engledew!
thanks for your contribution! 
it looks really great.
I'm testing it now

File: pkg/report/template.go [hunk 1]
```diff
@@ -46,6 +46,16 @@ func NewTemplateWriter(output io.Writer, outputTemplate string) (*TemplateWriter
 		}
 		return escaped.String()
 	}
+	templateFuncMap["makeRuleMap"] = func() map[string]int {
+		return make(map[string]int)
+	}
+	templateFuncMap["indexRule"] = func(rules map[string]int, vulnerabilityID string) bool {
+		if _, ok := rules[vulnerabilityID]; !ok {
+			rules[vulnerabilityID] = len(rules) + 1
```
Comment (afdesk) [inline pkg/report/template.go:54]: [sarif validator](https://sarifweb.azurewebsites.net/Validation) shows an error:
```
SARIF1009: runs[0].results[0]: This 'result' object contains a property 'ruleIndex' with value 1,
but 'runs[0].tool.driver.rules' has fewer than 2 elements. 
An index-valued properties must be valid for the array that it refers to.
```
it seems that we should remove `+1`, right?

Comment (simon-engledew) [reply pkg/report/template.go:54]: Ah yes, my bad! 😓

**Comment by simon-engledew** [comment]
Does what I have done make sense for Trivy?

I guess my main concern with this approach is that the rules would have result-specific information in them because of the way Trivy has a 1-1 mapping between results and rules at the moment.

**Comment by simon-engledew** [comment]
Hmm 🤔 Looking at the sample data more carefully, I think this is going to require a bit more work to resolve.

I've altered the rule metadata so that it is also stable and applicable to all results. I've had to remove the help entirely as it contained information specific to a single result.

If you want to reinstate this information it should really live inside results, e.g: the message field.

**Comment by knqyf263** [comment]
Hi @simon-engledew, I'm so sorry to be late. We're now putting the highest priority on it. Could you fix the test?

**Comment by simon-engledew** [comment]
Would it make sense to create a new PR based on this one? I feel like I am making decisions that would be better made by someone with more context about how Trivy works. 🙇

**Review by afdesk — APPROVED** [review]

File: contrib/sarif.tpl [hunk 2]
```diff
@@ -14,42 +15,39 @@
...
+                  {{- if $t_first -}}
+                    {{- $t_first = false -}}
+                  {{ else -}}
+                    ,
+                  {{- end }}
+                {
+                  "id": {{ .VulnerabilityID | toJson }},
+                  "name": "{{ toSarifRuleName $vulnerabilityType }}",
+                  "shortDescription": {
+                    "text": {{ .VulnerabilityID | toJson }},
```
Comment (afdesk) [inline contrib/sarif.tpl:28]: the trailing comma is wrong

**Review by afdesk — CHANGES_REQUESTED** [review]

**Review by afdesk — APPROVED** [review]

**Review by afdesk — APPROVED** [review]

**Review by afdesk — APPROVED** [review]

**Review by afdesk — APPROVED** [review]

**Review by afdesk — APPROVED** [review]

**Review by afdesk — APPROVED** [review]

**Comment by simon-engledew** [comment]
Ok `go test -v -tags=integration -run TestClientServerWithTemplate ./integration/...` now passes. 🙇  Sorry, I've been busy with other stuff all day and this is the first time I've been able to sit down and try and get the integration tests passing. 😞 

It's the end of the day for me now so if this doesn't do it I am more than happy if you want to cherry pick my commits, make a PR that is not based on my fork and run with it without me.

**Review by knqyf263 — APPROVED** [review]
Thanks a lot! We will improve our Sarif generation later.

**Comment by afdesk** [comment]
@simon-engledew thanks for your time and effort!

**Comment by simon-engledew** [comment]
Thanks for bearing with me as I repeatedly tried and failed to fix the tests @afdesk 😅❤️🙏

**Comment by afdesk** [comment]
@simon-engledew You spent a lot of time. thanks again

**Comment by afdesk** [comment]
Hi @simon-engledew!
could you help us again?
several parts have been removed from the sarif template: `help` block and `PkgName`.

It seems that was a mistake, right?

https://github.com/aquasecurity/trivy/blob/eaf2da20a690874570509e9924e2d5f1680c813d/contrib/sarif.tpl#L40-L49

**Comment by simon-engledew** [comment]
Ah, that's what I was refering to in this comment: https://github.com/aquasecurity/trivy/pull/1230#issuecomment-966363669

The rule has to be applicable to every rule with the same sarif identifier. It should contain some general text talking about the vulnerability and then result specific details should be included in the result object.

**Comment by afdesk** [comment]
> Ah, that's what I was refering to in this comment: [#1230 (comment)](https://github.com/aquasecurity/trivy/pull/1230#issuecomment-966363669)
> 
> The rule help has to be applicable to every rule with the same sarif identifier. It should contain some general text talking about the vulnerability and then result specific details should be included in the result object.

@simon-engledew thanks a lot. I missed this comment :(

**Pull request merged**